Panic when jaeger-tags value is set to env var without default value

When using the `--jaeger-tags` flag with a value being set by env vars, the agent crashes when no default value is specified. Example:

```
--jaeger.tags=key=value,envVar1=${envKey1:defaultVal1},envVar2=${envKey2:defaultVal2},envVar3=${envKey3}
```

The Jaeger Agent will crash when parsing the value for `envKey3`, as it expects a default value to exist:

```
panic: runtime error: index out of range

goroutine 1 [running]:
github.com/jaegertracing/jaeger/cmd/agent/app/reporter.parseAgentTags(0x7ffd2b4197bb, 0x69, 0xb)
	/home/travis/gopath/src/github.com/jaegertracing/jaeger/cmd/agent/app/reporter/flags.go:72 +0x430
github.com/jaegertracing/jaeger/cmd/agent/app/reporter.(*Options).InitFromViper(0xc000127af0, 0xc0000b2900, 0x5)
	/home/travis/gopath/src/github.com/jaegertracing/jaeger/cmd/agent/app/reporter/flags.go:55 +0x9e
main.main.func1(0xc0001d0a00, 0xc00016bc20, 0x0, 0x3, 0x0, 0x0)
	/home/travis/gopath/src/github.com/jaegertracing/jaeger/cmd/agent/main.go:59 +0x121
github.com/jaegertracing/jaeger/vendor/github.com/spf13/cobra.(*Command).execute(0xc0001d0a00, 0xc000092010, 0x3, 0x3, 0xc0001d0a00, 0xc000092010)
	/home/travis/gopath/src/github.com/jaegertracing/jaeger/vendor/github.com/spf13/cobra/command.go:762 +0x465
github.com/jaegertracing/jaeger/vendor/github.com/spf13/cobra.(*Command).ExecuteC(0xc0001d0a00, 0xc0001d0a00, 0xc000127f60, 0x5)
	/home/travis/gopath/src/github.com/jaegertracing/jaeger/vendor/github.com/spf13/cobra/command.go:852 +0x2ec
github.com/jaegertracing/jaeger/vendor/github.com/spf13/cobra.(*Command).Execute(...)
	/home/travis/gopath/src/github.com/jaegertracing/jaeger/vendor/github.com/spf13/cobra/command.go:800
main.main()
	/home/travis/gopath/src/github.com/jaegertracing/jaeger/cmd/agent/main.go:101 +0x3e8
```

--- a/cmd/agent/app/reporter/flags.go
+++ b/cmd/agent/app/reporter/flags.go
@@ -69,6 +69,11 @@ func parseAgentTags(agentTags string) map[string]string {
 
 		if strings.HasPrefix(v, "${") && strings.HasSuffix(v, "}") {
 			ed := strings.SplitN(string(v[2:len(v)-1]), ":", 2)
+			if (len(ed) == 1) {
+				// no default value specified, set to empty
+				ed = append(ed, "")
+			}
+
 			e, d := ed[0], ed[1]
 			v = os.Getenv(e)
 			if v == "" && d != "" {

--- a/cmd/agent/app/reporter/flags.go
+++ b/cmd/agent/app/reporter/flags.go
@@ -68,9 +68,12 @@ func parseAgentTags(agentTags string) map[string]string {
 		k, v := strings.TrimSpace(kv[0]), strings.TrimSpace(kv[1])
 
 		if strings.HasPrefix(v, "${") && strings.HasSuffix(v, "}") {
+			skipWhenEmpty := false
+
 			ed := strings.SplitN(string(v[2:len(v)-1]), ":", 2)
-			if (len(ed) == 1) {
+			if len(ed) == 1 {
 				// no default value specified, set to empty
+				skipWhenEmpty = true
 				ed = append(ed, "")
 			}
 
@@ -78,6 +81,11 @@ func parseAgentTags(agentTags string) map[string]string {
 			v = os.Getenv(e)
 			if v == "" && d != "" {
 				v = d
+			}
+
+			// no value is set, skip this entry
+			if v == "" && skipWhenEmpty {
+				continue
 			}
 		}
 

--- a/cmd/agent/app/reporter/flags_test.go
+++ b/cmd/agent/app/reporter/flags_test.go
@@ -16,6 +16,7 @@ package reporter
 
 import (
 	"flag"
+	"fmt"
 	"os"
 	"testing"
 
@@ -52,22 +53,36 @@ func TestBindFlags(t *testing.T) {
 	command.PersistentFlags().AddGoFlagSet(flags)
 	v.BindPFlags(command.PersistentFlags())
 
+	jaegerTags := fmt.Sprintf("%s,%s,%s,%s,%s,%s",
+		"key=value",
+		"envVar1=${envKey1:defaultVal1}",
+		"envVar2=${envKey2:defaultVal2}",
+		"envVar3=${envKey3}",
+		"envVar4=${envKey4}",
+		"envVar5=${envVar5:}",
+	)
+
 	err := command.ParseFlags([]string{
 		"--reporter.type=grpc",
-		"--jaeger.tags=key=value,envVar1=${envKey1:defaultVal1},envVar2=${envKey2:defaultVal2},envVar3=${envKey3}",
+		"--jaeger.tags=" + jaegerTags,
 	})
 	require.NoError(t, err)
 
 	b := &Options{}
 	os.Setenv("envKey1", "envVal1")
 	defer os.Unsetenv("envKey1")
+
+	os.Setenv("envKey4", "envVal4")
+	defer os.Unsetenv("envKey4")
+
 	b.InitFromViper(v)
 
 	expectedTags := map[string]string{
 		"key":     "value",
 		"envVar1": "envVal1",
 		"envVar2": "defaultVal2",
-		"envVar3": "",
+		"envVar4": "envVal4",
+		"envVar5": "",
 	}
 
 	assert.Equal(t, Type("grpc"), b.ReporterType)

--- a/cmd/agent/app/reporter/flags_test.go
+++ b/cmd/agent/app/reporter/flags_test.go
@@ -54,21 +54,22 @@ func TestBindFlags(t *testing.T) {
 
 	err := command.ParseFlags([]string{
 		"--reporter.type=grpc",
-		"--jaeger.tags=key=value,envVar1=${envKey1:defaultVal1},envVar2=${envKey2:defaultVal2}",
+		"--jaeger.tags=key=value,envVar1=${envKey1:defaultVal1},envVar2=${envKey2:defaultVal2},envVar3=${envKey3}",
 	})
 	require.NoError(t, err)
 
 	b := &Options{}
 	os.Setenv("envKey1", "envVal1")
+	defer os.Unsetenv("envKey1")
 	b.InitFromViper(v)
 
 	expectedTags := map[string]string{
 		"key":     "value",
 		"envVar1": "envVal1",
 		"envVar2": "defaultVal2",
+		"envVar3": "",
 	}
 
 	assert.Equal(t, Type("grpc"), b.ReporterType)
 	assert.Equal(t, expectedTags, b.AgentTags)
-	os.Unsetenv("envKey1")
 }

